### ISSUE TITLE
ssh-audit failing on some cisco devices

### ISSUE THREAD

**Issue by noblemtw** [jtesta/ssh-audit#90]
The later versions of ssh-audit are failing with the below error when trying to scan some Cisco devices. The ones I know of are running Cisco IOS XE Software, Version 16.10.01, although it's possible other versions have similar issues. The rather old version 1.7 worked, so it's something the more recent development has caused...

Unfortunately, due to firewalls and ACLs, I don't have any devices you would be able to directly test against but I'm happy to provide any necessary info to try and resolve the problem

```
Traceback (most recent call last):
  File "/opt/VMED/bin/ssh-audit.py", line 16, in <module>
    exit_code = main()
  File "/opt/VMED/bin/src/ssh_audit/ssh_audit.py", line 972, in main
    return audit(aconf)
  File "/opt/VMED/bin/src/ssh_audit/ssh_audit.py", line 835, in audit
    HostKeyTest.run(s, kex)
  File "/opt/VMED/bin/src/ssh_audit/hostkeytest.py", line 85, in run
    HostKeyTest.perform_test(s, server_kex, kex_str, kex_group, HostKeyTest.HOST_KEY_TYPES)
  File "/opt/VMED/bin/src/ssh_audit/hostkeytest.py", line 123, in perform_test
    SSH2_Kex.parse(payload)
  File "/opt/VMED/bin/src/ssh_audit/ssh2_kex.py", line 121, in parse
    kex_algs = buf.read_list()
  File "/opt/VMED/bin/src/ssh_audit/readbuf.py", line 57, in read_list
    list_size = self.read_int()
  File "/opt/VMED/bin/src/ssh_audit/readbuf.py", line 53, in read_int
    v = struct.unpack('>I', self.read(4))[0]  # type: int
struct.error: unpack requires a buffer of 4 bytes
```

**Comment by jtesta** [comment]
Thanks for reporting this!  And sorry for the late response.

My guess is that the service stops responding to new connections during
the testing (connection throttling?), causing the hostkey checks to
fail.  But I'd rather not guess, if at all possible.  Could I get a
PCAP file showing what happens when you run ssh-audit against this
Cisco device?  If so, you can e-mail it to me directly so it's not made
public (jtesta at-sign positronsecurity dot com).

Otherwise, the best I can do is make ssh-audit skip hostkey tests once
the first error is encountered...

-- 
Joseph S. Testa II
Founder & Principal Security Consultant
Positron Security

**Comment by noblemtw** [comment]
Hi Joe,

Many thanks for looking into this and don't worry about the delayed
response, I imagine you have a day job like the rest of us.

I've attached two packet captures. The first one (ssh2.dmp) is from 2.3.1,
however while running the first I noticed some errors at the router end
(below) and I think the connection was just being dropped due to
client/server failing to negotiate a suitable kex. I think this is possibly
a recent thing as I don't believe it was doing this before, although it
never occurred to me to run a packet capture when I last tried to debug the
issue

Jan 20 10:00:29 GMT: %SSH-3-NO_MATCH: No matching kex algorithm found:
client curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256
server diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1

The second capture (ssh3.dmp) is taken after adding
diffie-hellman-group14-sha1  in to the key_exchange list in *ssh_socket.py.
*I've just noticed the comment in there that says you're emulating OpenSSH
8.2. Cisco's SSH support seems to be rather poor and I've yet to see a
device which I don't need workarounds to access, although this could just
be our network...

Anyway, even with the extra Kex added, it's still failing, although now
with a slightly different error in the router logs:

*Jan 20 10:14:04 GMT: %SSH-4-SSH2_UNEXPECTED_MSG: Unexpected message type
has arrived. Terminating the connection from 62.253.167.162*

Hopefully there's enough info in the two packet captures to figure out
what's failing and resolve the issue. It would be great to run ssh-audit
against all our Cisco estate and highlight how bad some of the settings are
:)

Regards

Matt

**Comment by jtesta** [comment]
Unfortunately, Github didn't pick up the attached files from your e-mail response.  Seems like you'll have to log into the web application to attach files...

**Comment by noblemtw** [comment]
Hopefully these work...

[ssh.dmp.gz](https://github.com/jtesta/ssh-audit/files/5843686/ssh.dmp.gz)
[ssh3.dmp.gz](https://github.com/jtesta/ssh-audit/files/5843687/ssh3.dmp.gz)

**Comment by jtesta** [comment]
> Hopefully these work...

Yep, they did.

I think I identified the problem, too.  A potential fix has been
checked into master.  Can you give it a try?  If it doesn't work,
please attach another packet capture, along with the full program
output.  Thanks!

**Comment by noblemtw** [comment]
Wow, that was quick! 

Running master against the problem device is now reporting back as expected so the fix looks good

`# general
(gen) banner: SSH-2.0-Cisco-1.25
(gen) software: Cisco IOS/PIX sshd 1.25
(gen) compatibility: OpenSSH 5.9-6.6, Dropbear SSH 2013.56+
(gen) compression: disabled

# key exchange algorithms
(kex) diffie-hellman-group-exchange-sha1 (2048-bit) -- [fail] removed (in server) since OpenSSH 6.7, unsafe algorithm
                                                    `- [warn] using weak hashing algorithm
                                                    `- [info] available since OpenSSH 2.3.0
(kex) diffie-hellman-group14-sha1         -- [warn] using weak hashing algorithm
                                          `- [info] available since OpenSSH 3.9, Dropbear SSH 0.53`

**Comment by jtesta** [comment]
> Wow, that was quick!
> 
> Running master against the problem device is now reporting back as
> expected so the fix looks good

Great to hear!
